File: src/etch/examples/benchmark/etch/PerfHelper.cs
using System;
using System.Collections.Generic;
using System.Text;

using Etch.Msg;
using Etch.Support;
using Etch.Transport;

namespace etch.examples.benchmark.etch
{
    public class PerfHelper
    {
        public static ListenerHandler newPerfListener( Pool queuedPool, Pool freePool )
        {
            return new MyListenerHandler( queuedPool, freePool );
        }

        public static PerfServer newPerfServer( TcpConnection conn,
            PerfClient client, Pool queued, Pool free )
        {
            DefaultDeliveryService<StubPerfClient> dds = 
                new MyDeliveryService( client, queued, free );

            DefaultDeliveryService<StubPerfClient> svc = 
                DefaultDeliveryService<StubPerfClient>.SetUp( dds, conn, new ValueFactoryPerf() );

            return new RemotePerfServer( svc );
        }


        public class MyListenerHandler : DefaultListenerHandler
        {
            Pool _free;
            Pool _pool;

            public MyListenerHandler( Pool pool, Pool free )
            {
                _free = free;
                _pool = pool;
            }

            protected override Etch.Transport.MessageHandler<MailboxMessageSource> NewMessageHandler()
            {
                return new MyMessageHandler( _pool, _free );
            }

            protected override ValueFactory NewValueFactory()
            {
                return new ValueFactoryPerf();
            }

            public override string ToString()
            {
                return "PerfServerListener";
            }
        }

        public class MyMessageHandler : DefaultMessageHandler<StubPerfServer>
        {
            Pool _free;
            Pool _pool;

            public MyMessageHandler( Pool pool, Pool free )
            {
                _free = free;
                _pool = pool;

            }

            protected override StubPerfServer NewStub( MailboxMessageSource src )
            {
             
                return new StubPerfServer( new ImplPerfServer(), _pool, _free );
            }

            public override string ToString()
            {
                return "PerfServerMessageHandler";
            }
        }

        public class MyDeliveryService : DefaultDeliveryService<StubPerfClient>
        {
            PerfClient _client;
            Pool _free;
            Pool _queued;

            public MyDeliveryService( MailboxMessageSource src )
                : base( src )
            { }

            public MyDeliveryService( PerfClient client, Pool queued, Pool free )
            {
                _client = client;
                _free = free;
                _queued = queued;
            }

            protected override StubPerfClient NewStub( MailboxMessageSource src )
            {
                return new StubPerfClient( _client, _queued, _free );
            }

            public override string ToString()
            {
                return null;
            }
        }
    }
}

File: src/etch/examples/benchmark/etch/PerfHelper.java
package etch.examples.benchmark.etch;

import metreos.core.io.ListenerHandler;
import metreos.core.io.TcpConnection;
import etch.bindings.java.msg.ValueFactory;
import etch.bindings.java.support.Pool;
import etch.bindings.java.transport.DefaultDeliveryService2;
import etch.bindings.java.transport.DefaultListenerHandler;
import etch.bindings.java.transport.DefaultMessageHandler;
import etch.bindings.java.transport.MailboxMessageSource;
import etch.bindings.java.transport.MessageHandler;

/**
 * Helper for constructing Perf transports.
 */
public class PerfHelper
{
	/**
	 * @param queuedPool
	 * @param freePool
	 * @return a ListenerHandler to use.
	 */
	public static ListenerHandler newPerfListener( final Pool queuedPool,
		final Pool freePool )
	{
		return new DefaultListenerHandler()
		{
			@Override
			protected MessageHandler<MailboxMessageSource> newMessageHandler()
			{
				return new DefaultMessageHandler<StubPerfServer>()
				{
					@Override
					protected StubPerfServer newStub( MailboxMessageSource src )
					{
						return new StubPerfServer( src, new ImplPerfServer(),
							queuedPool, freePool );
					}
	
					@Override
					public String toString()
					{
						return "PerfServerMessageHandler";
					}
				};
			}
	
			@Override
			protected ValueFactory newValueFactory()
			{
				return new ValueFactoryPerf();
			}
	
			@Override
			public String toString()
			{
				return "PerfServerListener";
			}
		};
	}

	/**
	 * @param c
	 * @param client
	 * @param queuedPool
	 * @param freePool
	 * @return a PerfServer to use.
	 */
	public static PerfServer newPerfServer( TcpConnection c,
		final PerfClient client, final Pool queuedPool, final Pool freePool )
	{
		DefaultDeliveryService2<StubPerfClient> dds =
			new DefaultDeliveryService2<StubPerfClient>()
		{
			@Override
			protected StubPerfClient newStub( MailboxMessageSource src )
			{
				return new StubPerfClient( src, client, queuedPool, freePool );
			}

			@Override
			public String toString()
			{
				// TODO Auto-generated method stub
				return null;
			}
		};
		
		DefaultDeliveryService2<StubPerfClient> svc = DefaultDeliveryService2.setUp( dds,
			c, new ValueFactoryPerf() );
		
		return new RemotePerfServer( svc );
	}
}
